#Algoritmos - Actividad Guiada 1

Nombre: Marina Riquelme Morell<br>
URL: https://colab.research.google.com/drive/1CpWD5Z7ddo-acXNehyIWrTDJs6MSiBHS?usp=sharing

http:https://github.com/Ma7R7/Algoritmos-Optimizaci-n-Marina-/blob/main/Algoritmos_AG1%20(1).ipynb


##Torres de Hanoi con Divide y vencerás

In [1]:
def Torres_Hanoi(N, desde, hasta):
  if N ==1 :
    print("Lleva la ficha " ,desde , " hasta ", hasta )

  else:
    #Torres_Hanoi(N-1, desde, 6-desde-hasta )
    Torres_Hanoi(N-1, desde, 6-desde-hasta )
    print("Lleva la ficha " ,desde , " hasta ", hasta )
    #Torres_Hanoi(N-1,6-desde-hasta, hasta )
    Torres_Hanoi(N-1, 6-desde-hasta  , hasta )


Torres_Hanoi(3, 1 , 3)

Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3


In [2]:
#Sucesión_de_Fibonacci
#https://es.wikipedia.org/wiki/Sucesi%C3%B3n_de_Fibonacci
#Calculo del termino n-simo de la suscesión de Fibonacci
def Fibonacci(N:int):
  if N < 2:
    return 1
  else:
    return Fibonacci(N-1)+Fibonacci(N-2)

Fibonacci(5)

8

##Devolución de cambio por técnica voraz

In [3]:
def cambio_monedas(N, SM):
  SOLUCION = [0]*len(SM)   #SOLUCION = [0,0,0,0,..]
  ValorAcumulado = 0

  for i,valor in enumerate(SM):
    monedas =  (N-ValorAcumulado)//valor
    SOLUCION[i] = monedas
    ValorAcumulado = ValorAcumulado + monedas*valor

    if ValorAcumulado == N:
      return SOLUCION


cambio_monedas(15,[25,10,5,1])

[0, 1, 1, 0]

## N-Reinas por técnica de vueta atrás


In [4]:
def escribe(S):
  n = len(S)
  for x in range(n):
    print("")
    for i in range(n):
      if S[i] == x+1:
        print(" X " , end="")
      else:
        print(" - ", end="")


def es_prometedora(SOLUCION,etapa):
  #print(SOLUCION)
  #Si la solución tiene dos valores iguales no es valida => Dos reinas en la misma fila
  for i in range(etapa+1):
    #print("El valor " + str(SOLUCION[i]) + " está " +  str(SOLUCION.count(SOLUCION[i])) + " veces")
    if SOLUCION.count(SOLUCION[i]) > 1:
      return False

    #Verifica las diagonales
    for j in range(i+1, etapa +1 ):
      #print("Comprobando diagonal de " + str(i) + " y " + str(j))
      if abs(i-j) == abs(SOLUCION[i]-SOLUCION[j]) : return False
  return True



def reinas(N, solucion=[], etapa=0):
  if len(solucion) == 0:
      solucion=[0 for i in range(N)]

  for i in range(1, N+1):
    solucion[etapa] = i

    if es_prometedora(solucion, etapa):
      if etapa == N-1:
        print(solucion)
        #escribe(solucion)
        print()
      else:
        reinas(N, solucion, etapa+1)
    else:
      None

    solucion[etapa] = 0

reinas(8)

[1, 5, 8, 6, 3, 7, 2, 4]

[1, 6, 8, 3, 7, 4, 2, 5]

[1, 7, 4, 6, 8, 2, 5, 3]

[1, 7, 5, 8, 2, 4, 6, 3]

[2, 4, 6, 8, 3, 1, 7, 5]

[2, 5, 7, 1, 3, 8, 6, 4]

[2, 5, 7, 4, 1, 8, 6, 3]

[2, 6, 1, 7, 4, 8, 3, 5]

[2, 6, 8, 3, 1, 4, 7, 5]

[2, 7, 3, 6, 8, 5, 1, 4]

[2, 7, 5, 8, 1, 4, 6, 3]

[2, 8, 6, 1, 3, 5, 7, 4]

[3, 1, 7, 5, 8, 2, 4, 6]

[3, 5, 2, 8, 1, 7, 4, 6]

[3, 5, 2, 8, 6, 4, 7, 1]

[3, 5, 7, 1, 4, 2, 8, 6]

[3, 5, 8, 4, 1, 7, 2, 6]

[3, 6, 2, 5, 8, 1, 7, 4]

[3, 6, 2, 7, 1, 4, 8, 5]

[3, 6, 2, 7, 5, 1, 8, 4]

[3, 6, 4, 1, 8, 5, 7, 2]

[3, 6, 4, 2, 8, 5, 7, 1]

[3, 6, 8, 1, 4, 7, 5, 2]

[3, 6, 8, 1, 5, 7, 2, 4]

[3, 6, 8, 2, 4, 1, 7, 5]

[3, 7, 2, 8, 5, 1, 4, 6]

[3, 7, 2, 8, 6, 4, 1, 5]

[3, 8, 4, 7, 1, 6, 2, 5]

[4, 1, 5, 8, 2, 7, 3, 6]

[4, 1, 5, 8, 6, 3, 7, 2]

[4, 2, 5, 8, 6, 1, 3, 7]

[4, 2, 7, 3, 6, 8, 1, 5]

[4, 2, 7, 3, 6, 8, 5, 1]

[4, 2, 7, 5, 1, 8, 6, 3]

[4, 2, 8, 5, 7, 1, 3, 6]

[4, 2, 8, 6, 1, 3, 5, 7]

[4, 6, 1, 5, 2, 8, 3, 7]

[4, 6, 8, 2, 7, 1, 3, 5]

[4, 6, 8, 3,

## Viaje por el rio. Programación dinámica

In [5]:
TARIFAS = [
[0,5,4,3,999,999,999],
[999,0,999,2,3,999,11],
[999,999, 0,1,999,4,10],
[999,999,999, 0,5,6,9],
[999,999, 999,999,0,999,4],
[999,999, 999,999,999,0,3],
[999,999,999,999,999,999,0]
]



################################################################
def Precios(TARIFAS):
################################################################
  #Total de Nodos
  N = len(TARIFAS[0])

  #Inicialización de la tabla de precios
  PRECIOS = [ [9999]*N for i in [9999]*N]
  RUTA = [ [""]*N for i in [""]*N]

  for i in range(0,N-1):
    RUTA[i][i] = i             #Para ir de i a i se "pasa por i"
    PRECIOS[i][i] = 0          #Para ir de i a i se se paga 0
    for j in range(i+1, N):
      MIN = TARIFAS[i][j]
      RUTA[i][j] = i

      for k in range(i, j):
        if PRECIOS[i][k] + TARIFAS[k][j] < MIN:
            MIN = min(MIN, PRECIOS[i][k] + TARIFAS[k][j] )
            RUTA[i][j] = k          #Anota que para ir de i a j hay que pasar por k
        PRECIOS[i][j] = MIN

  return PRECIOS,RUTA
################################################################


PRECIOS,RUTA = Precios(TARIFAS)
#print(PRECIOS[0][6])

print("PRECIOS")
for i in range(len(TARIFAS)):
  print(PRECIOS[i])

print("\nRUTA")
for i in range(len(TARIFAS)):
  print(RUTA[i])

#Determinar la ruta con Recursividad
def calcular_ruta(RUTA, desde, hasta):
  if desde == hasta:
    #print("Ir a :" + str(desde))
    return ""
  else:
    return str(calcular_ruta( RUTA, desde, RUTA[desde][hasta])) +  \
                ',' + \
                str(RUTA[desde][hasta] \
              )

print("\nLa ruta es:")
calcular_ruta(RUTA, 0,6)

PRECIOS
[0, 5, 4, 3, 8, 8, 11]
[9999, 0, 999, 2, 3, 8, 7]
[9999, 9999, 0, 1, 6, 4, 7]
[9999, 9999, 9999, 0, 5, 6, 9]
[9999, 9999, 9999, 9999, 0, 999, 4]
[9999, 9999, 9999, 9999, 9999, 0, 3]
[9999, 9999, 9999, 9999, 9999, 9999, 9999]

RUTA
[0, 0, 0, 0, 1, 2, 5]
['', 1, 1, 1, 1, 3, 4]
['', '', 2, 2, 3, 2, 5]
['', '', '', 3, 3, 3, 3]
['', '', '', '', 4, 4, 4]
['', '', '', '', '', 5, 5]
['', '', '', '', '', '', '']

La ruta es:


',0,2,5'

## Ejercicio adicional propuesto: encontrar los dos puntos más cercanos.


**Planteamiento del problema**

El problema exige localizar, dentro de un conjunto de puntos espaciales, el par que presenta la menor distancia métrica entre sí.

La resolución se aborda mediante las siguientes etapas:

* Búsqueda de la solución en un espacio unidimensional empleando un algoritmo inicial de fuerza bruta, el cual evalúa de forma exhaustiva todas las combinaciones posibles, junto con el análisis de su complejidad computacional.

* Optimización de la búsqueda unidimensional mediante la implementación del paradigma de divide y vencerás con el fin de reducir el coste temporal, evaluando nuevamente la mejora en la complejidad teórica.

* Extensión de la lógica geométrica y del diseño algorítmico de divide y vencerás para resolver de manera análoga el problema en espacios de dos y tres dimensiones.

* Generación de conjuntos de datos aleatorios, conformados por listas de mil puntos espaciales, para poner a prueba el rendimiento, la exactitud y la escalabilidad computacional de los métodos desarrollados en cada una de las dimensiones.

In [9]:
import math
import random
import time

# Generación de los conjuntos de datos aleatorios.
LISTA_1D = [random.randrange(1, 10000) for x in range(1000)]
LISTA_2D = [(random.randrange(1, 10000), random.randrange(1, 10000)) for x in range(1000)]
LISTA_3D = [(random.randrange(1, 10000), random.randrange(1, 10000), random.randrange(1, 10000)) for x in range(1000)]

def calcular_distancia_1d(punto_1, punto_2):
    """
    Calcula la distancia entre dos puntos en una dimensión.
    """
    return abs(punto_1 - punto_2)

def calcular_distancia_2d(punto_1, punto_2):
    """
    Calcula la distancia euclídea entre dos puntos en dos dimensiones.
    """
    return math.sqrt((punto_1[0] - punto_2[0])**2 + (punto_1[1] - punto_2[1])**2)

def calcular_distancia_3d(punto_1, punto_2):
    """
    Calcula la distancia euclídea entre dos puntos en tres dimensiones.
    """
    return math.sqrt((punto_1[0] - punto_2[0])**2 + (punto_1[1] - punto_2[1])**2 + (punto_1[2] - punto_2[2])**2)

#################################################################################################################
#Fuerza bruta
#################################################################################################################

def fuerza_bruta(puntos, funcion_distancia):
    """
    Encuentra el par más cercano evaluando todas las combinaciones.
    Es aplicable a cualquier dimensión pasando la función de distancia correspondiente.
    """
    distancia_minima = float('inf')
    mejor_par = None
    numero_puntos = len(puntos)

    for i in range(numero_puntos):
        for j in range(i + 1, numero_puntos):
            distancia_actual = funcion_distancia(puntos[i], puntos[j])
            if distancia_actual < distancia_minima:
                distancia_minima = distancia_actual
                mejor_par = (puntos[i], puntos[j])

    return mejor_par, distancia_minima

#################################################################################################################
#Divide y vencerás
#################################################################################################################

def divide_y_venceras_1d(puntos):
    """
    Optimiza la búsqueda en una dimensión mediante una ordenación previa.
    """
    puntos_ordenados = sorted(puntos)
    distancia_minima = float('inf')
    mejor_par = None

    for i in range(len(puntos_ordenados) - 1):
        distancia_actual = calcular_distancia_1d(puntos_ordenados[i], puntos_ordenados[i+1])
        if distancia_actual < distancia_minima:
            distancia_minima = distancia_actual
            mejor_par = (puntos_ordenados[i], puntos_ordenados[i+1])

    return mejor_par, distancia_minima

def divide_y_venceras_2d(puntos_x, puntos_y):
    """
    Se resuelve el problema bidimensional dividiendo el plano recursivamente.
    """
    numero_puntos = len(puntos_x)

    if numero_puntos <= 3:
        return fuerza_bruta(puntos_x, calcular_distancia_2d)

    mitad = numero_puntos // 2
    punto_central = puntos_x[mitad]

    puntos_izq_x = puntos_x[:mitad]
    puntos_der_x = puntos_x[mitad:]

    puntos_izq_y = [p for p in puntos_y if p[0] <= punto_central[0]]
    puntos_der_y = [p for p in puntos_y if p[0] > punto_central[0]]

    par_izq, dist_izq = divide_y_venceras_2d(puntos_izq_x, puntos_izq_y)
    par_der, dist_der = divide_y_venceras_2d(puntos_der_x, puntos_der_y)

    if dist_izq < dist_der:
        distancia_minima = dist_izq
        mejor_par = par_izq
    else:
        distancia_minima = dist_der
        mejor_par = par_der

    franja = [p for p in puntos_y if abs(p[0] - punto_central[0]) < distancia_minima]

    for i in range(len(franja)):
        for j in range(i + 1, min(i + 8, len(franja))):
            distancia_actual = calcular_distancia_2d(franja[i], franja[j])
            if distancia_actual < distancia_minima:
                distancia_minima = distancia_actual
                mejor_par = (franja[i], franja[j])

    return mejor_par, distancia_minima

def divide_y_venceras_3d(puntos_ordenados_x):
    """
    Se resuelve el problema tridimensional dividiendo el espacio mediante planos.
    """
    numero_puntos = len(puntos_ordenados_x)

    if numero_puntos <= 3:
        return fuerza_bruta(puntos_ordenados_x, calcular_distancia_3d)

    mitad = numero_puntos // 2
    punto_central = puntos_ordenados_x[mitad]

    puntos_izq = puntos_ordenados_x[:mitad]
    puntos_der = puntos_ordenados_x[mitad:]

    par_izq, dist_izq = divide_y_venceras_3d(puntos_izq)
    par_der, dist_der = divide_y_venceras_3d(puntos_der)

    if dist_izq < dist_der:
        distancia_minima = dist_izq
        mejor_par = par_izq
    else:
        distancia_minima = dist_der
        mejor_par = par_der

    franja = [p for p in puntos_ordenados_x if abs(p[0] - punto_central[0]) < distancia_minima]
    franja.sort(key=lambda p: p[1])

    for i in range(len(franja)):
        for j in range(i + 1, len(franja)):
            if franja[j][1] - franja[i][1] >= distancia_minima:
                break

            if abs(franja[j][2] - franja[i][2]) >= distancia_minima:
                continue

            distancia_actual = calcular_distancia_3d(franja[i], franja[j])
            if distancia_actual < distancia_minima:
                distancia_minima = distancia_actual
                mejor_par = (franja[i], franja[j])

    return mejor_par, distancia_minima
#################################################################################################################
# Ejecutamos y medimos los tiempos
#################################################################################################################


if __name__ == "__main__":
    print("--- Análisis en 1D ---")
    inicio = time.time()
    par_fb_1d, dist_fb_1d = fuerza_bruta(LISTA_1D, calcular_distancia_1d)
    tiempo_fb_1d = time.time() - inicio

    inicio = time.time()
    par_dv_1d, dist_dv_1d = divide_y_venceras_1d(LISTA_1D)
    tiempo_dv_1d = time.time() - inicio

    print(f"Fuerza bruta -> Par: {par_fb_1d}, Distancia: {dist_fb_1d}, Tiempo: {tiempo_fb_1d*1000:.2f} ms")
    print(f"Divide y vencerás -> Par: {par_dv_1d}, Distancia: {dist_dv_1d}, Tiempo: {tiempo_dv_1d*1000:.2f} ms\n")

    print("--- Análisis en 2D ---")
    inicio = time.time()
    par_fb_2d, dist_fb_2d = fuerza_bruta(LISTA_2D, calcular_distancia_2d)
    tiempo_fb_2d = time.time() - inicio

    lista_2d_x = sorted(LISTA_2D, key=lambda p: p[0])
    lista_2d_y = sorted(LISTA_2D, key=lambda p: p[1])

    inicio = time.time()
    par_dv_2d, dist_dv_2d = divide_y_venceras_2d(lista_2d_x, lista_2d_y)
    tiempo_dv_2d = time.time() - inicio

    print(f"Fuerza bruta -> Par: {par_fb_2d}, Distancia: {dist_fb_2d:.4f}, Tiempo: {tiempo_fb_2d*1000:.2f} ms")
    print(f"Divide y vencerás -> Par: {par_dv_2d}, Distancia: {dist_dv_2d:.4f}, Tiempo: {tiempo_dv_2d*1000:.2f} ms\n")

    print("--- Análisis en 3D ---")
    inicio = time.time()
    par_fb_3d, dist_fb_3d = fuerza_bruta(LISTA_3D, calcular_distancia_3d)
    tiempo_fb_3d = time.time() - inicio

    puntos_x_3d = sorted(LISTA_3D, key=lambda p: p[0])

    inicio = time.time()
    par_dv_3d, dist_dv_3d = divide_y_venceras_3d(puntos_x_3d)
    tiempo_dv_3d = time.time() - inicio

    print(f"Fuerza bruta -> Par: {par_fb_3d}, Distancia: {dist_fb_3d:.4f}, Tiempo: {tiempo_fb_3d*1000:.2f} ms")
    print(f"Divide y vencerás -> Par: {par_dv_3d}, Distancia: {dist_dv_3d:.4f}, Tiempo: {tiempo_dv_3d*1000:.2f} ms")

--- Análisis en 1D ---
Fuerza bruta -> Par: (1930, 1930), Distancia: 0, Tiempo: 145.90 ms
Divide y vencerás -> Par: (195, 195), Distancia: 0, Tiempo: 0.48 ms

--- Análisis en 2D ---
Fuerza bruta -> Par: ((686, 2382), (679, 2380)), Distancia: 7.2801, Tiempo: 317.98 ms
Divide y vencerás -> Par: ((679, 2380), (686, 2382)), Distancia: 7.2801, Tiempo: 12.05 ms

--- Análisis en 3D ---
Fuerza bruta -> Par: ((8185, 2628, 203), (8205, 2578, 161)), Distancia: 68.2935, Tiempo: 394.97 ms
Divide y vencerás -> Par: ((8205, 2578, 161), (8185, 2628, 203)), Distancia: 68.2935, Tiempo: 4.89 ms



**Análisis de la complejidad matemática**

* El enfoque de fuerza bruta presenta una complejidad temporal acotada por $O(n^2)$ independientemente de la dimensionalidad del problema.

* La estrategia de divide y vencerás en una dimensión establece una complejidad general de $O(n \log n)$. El esfuerzo computacional principal es absorbido por la rutina de ordenación inicial. Tras ella, la comprobación de distancias mínimas se resuelve mediante un barrido lineal de coste $O(n)$.

* Las implementaciones bidimensional y tridimensional bajo el paradigma de partición logran mantener la cota de complejidad teórica en $O(n \log n)$.

**Interpretación de los resultados empíricos**

* El ensayo unidimensional revela una mejora  en el rendimiento al reducir el tiempo de procesamiento desde 145.90 milisegundos con la fuerza bruta hasta apenas 0.48 milisegundos mediante la técnica de divide y vencerás.

* El salto al plano bidimensional penaliza a la fuerza bruta, elevando su tiempo a 317.98 milisegundos. Por su parte, la aproximación recursiva lo resuelve en 12.05 milisegundos.

* El escenario tridimensional pone de manifiesto la escalabilidad real de la solución optimizada. Mientras la técnica exhaustiva escala hasta rozar los 394.97 milisegundos, el enfoque de divide y vencerás procesa el volumen espacial en tan solo 4.89 milisegundos gracias a la poda sistemática de evaluaciones basada en el ordenamiento interno por dimensiones.

**Conclusiones**

- Se han obtenido resultados idénticos con los dos métodos evaluados, demostrando que la lógica de partición del espacio funciona de manera correcta y no pierde precisión matemática frente a la evaluación exhaustiva.

- Se constata que aumentar el número de dimensiones retrasa considerablemente el tiempo de cálculo si empleamos la fuerza bruta, lo que demuestra que esta solución inicial es inútil cuando manejamos conjuntos de datos muy grandes.

- Se ha demostrado que implementar el paradigma de divide y vencerás constituye una alternativa viable para procesar sistemas con elevada densidad de puntos de manera rápida y eficiente.

Bibliografía

Reyero Díez, R. (2026). *Algoritmos de optimización* [Manual de asignatura]. Universidad Internacional de Valencia.
